ssl/statem/statem_lib.c

```diff
@@ -2403,3 +2403,149 @@ int tls13_restore_handshake_digest_for_pha(SSL *s)
+            || !WPACKET_get_total_written(&tmppkt, &length)
+            || !WPACKET_put_bytes_u24(pkt, length))
+        goto err;
+
+    /*
+     * Uncompressibility expansion:
+     * Brotli: per RFC7932: N + 5 + 3 * (N >> 16)
+     * ZLIB: N + 11 + 5 * (N >> 14)
+     * ZSTD: N + 4 + 14  \+ 3 * (N >> 17) + 4
+     */
+    switch (s->ext.compress_certificate_rx) {
+# ifdef ZLIB
+    case TLSEXT_comp_cert_zlib:
+        max_length = length + 11 + 5 * (length >> 14);
+        method = COMP_zlib();
+        break;
+# endif
+# ifdef BROTLI
+    case TLSEXT_comp_cert_brotli:
+        max_length = length + 5 + 3 * (length >> 16);
+        method = COMP_brotli_oneshot();
+        break;
+# endif
+# ifdef ZSTD
+    case TLSEXT_comp_cert_zstd:
+        max_length = length + 22 + 3 * (length >> 17);
+        method = COMP_zstd_oneshot();
+        break;
+# endif
+    default:
+        goto err;
+    }
+
+    if (!WPACKET_reserve_bytes(pkt, max_length+3, NULL)
+            || (comp = COMP_CTX_new(method)) == NULL)
+        goto err;
+
+    comp_len = COMP_compress_block(comp, WPACKET_get_curr(pkt)+3, max_length,
+                                   (unsigned char*)buf->data, length);
+    if (comp_len <= 0)
+        goto err;
+
+    /* Put length then move write pointer past certificate */
+    if (!WPACKET_put_bytes_u24(pkt, comp_len)
+            || !WPACKET_advance_write(pkt, comp_len))
+        goto err;
+
+    ret = 1;
+    goto out;
+ err:
+    SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+ out:
+    BUF_MEM_free(buf);
+    COMP_CTX_free(comp);
+
+    return ret;
+}
+
+MSG_PROCESS_RETURN tls13_process_compressed_certificate(SSL *s, PACKET *pkt, PACKET *tmppkt, BUF_MEM **buf)
+{
+
+    MSG_PROCESS_RETURN ret = MSG_PROCESS_ERROR;
+    unsigned int comp_alg;
+    COMP_METHOD *method = NULL;
+    COMP_CTX *comp = NULL;
+    size_t expected_length;
+    size_t comp_length;
+
+    if (!PACKET_get_net_2(pkt, &comp_alg)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    switch (comp_alg) {
+# ifdef ZLIB
+    case TLSEXT_comp_cert_zlib:
+        method = COMP_zlib();
+        break;
+# endif
+# ifdef BROTLI
+    case TLSEXT_comp_cert_brotli:
+        method = COMP_brotli_oneshot();
+        break;
+# endif
+# ifdef ZSTD
+    case TLSEXT_comp_cert_zstd:
+        method = COMP_zstd_oneshot();
+        break;
+# endif
+    default:
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_R_BAD_COMPRESSION_ALGORITHM);
+        goto err;
+    }
+
+    if (!PACKET_get_net_3_len(pkt, &expected_length)
+        || !PACKET_get_net_3_len(pkt, &comp_length)
+        || PACKET_remaining(pkt) != comp_length
+        || (*buf = BUF_MEM_new()) == NULL
+        || !BUF_MEM_grow(*buf, expected_length)
+        || !PACKET_buf_init(tmppkt, (unsigned char*)(*buf)->data, expected_length)
+        || (comp = COMP_CTX_new(method)) == NULL
+        || COMP_expand_block(comp, (unsigned char*)(*buf)->data, expected_length,
```

Comment:
> uncompressed_length:  The length of the Certificate message once it
>       is uncompressed.  If, after decompression, the specified length
>       does not match the actual length, the party receiving the invalid
>       message MUST abort the connection with the "bad_certificate"
>       alert.

Maybe **SSL_R_SSLV3_ALERT_BAD_CERTIFICATE** will be better here?

Reply:
That might be an appropriate alert.

Reply:
Based on a recent reading of RFC8879, this alert should be used here, and in a few other places.